### PR TITLE
Rules priorities

### PR DESCRIPTION
**Pull request by igor-lu**
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/Natural-Intelligence/rule-bos/pull/29%23discussion_r91548723%22%2C%20%22https%3A//github.com/Natural-Intelligence/rule-bos/pull/29%23discussion_r91549716%22%2C%20%22https%3A//github.com/Natural-Intelligence/rule-bos/pull/29%23discussion_r91549544%22%2C%20%22https%3A//github.com/Natural-Intelligence/rule-bos/pull/29%23discussion_r91856079%22%5D%2C%20%22comments%22%3A%20%7B%22Pull%20ba4ee2925514270db1a37df0443cb78d2c29a2bd%20tests/end2end/umbrella/rule-priority-controller-tests.js%2056%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/Natural-Intelligence/rule-bos/pull/29%23discussion_r91549544%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Need%20to%20add%20e2e%20tests%20%28involves%20rule%20%26%20rule%20priority%20unit%20tests%29%20that%3A%5Cr%5Cn1.%20create%20a%20rule%5Cr%5Cn2.%20create%20a%20rule%20prioirity%5Cr%5Cn3.%20create%20a%20rule%5Cr%5Cn4.%20update%20rule%20prioirity%5Cr%5Cn5.%20verify%20order%5Cr%5Cn6.%20delete%20rule%5Cr%5Cn7.%20update%20priority%5Cr%5Cn8.%20verify%20order%5Cr%5Cn%5Cr%5CnEtc.%22%2C%20%22created_at%22%3A%20%222016-12-08T16%3A27%3A56Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/14094514%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/moranzabatani%22%7D%7D%2C%20%7B%22body%22%3A%20%22rules%20and%20rules%20priorities%20are%20fully%20independent%20in%20server%20side%22%2C%20%22created_at%22%3A%20%222016-12-11T13%3A42%3A32Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/19372366%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/igor-lu%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20tests/end2end/umbrella/rule-priority-controller-tests.js%3AL1-316%22%7D%2C%20%22Pull%20ba4ee2925514270db1a37df0443cb78d2c29a2bd%20tests/end2end/umbrella/rule-priority-controller-tests.js%2083%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/Natural-Intelligence/rule-bos/pull/29%23discussion_r91549716%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Need%20to%20add%20e2e%20tests%20%28involves%20experiment%20%26%20experiment%20priority%20unit%20tests%29%20-%20the%20same%20as%20above%22%2C%20%22created_at%22%3A%20%222016-12-08T16%3A28%3A38Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/14094514%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/moranzabatani%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20tests/end2end/umbrella/rule-priority-controller-tests.js%3AL1-316%22%7D%2C%20%22Pull%20ba4ee2925514270db1a37df0443cb78d2c29a2bd%20tests/end2end/boost/experiment-controller-tests.js%206%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/Natural-Intelligence/rule-bos/pull/29%23discussion_r91548723%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22why%20remove%20all%20others%20parameter%20from%20context%3F%22%2C%20%22created_at%22%3A%20%222016-12-08T16%3A24%3A23Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/14094514%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/moranzabatani%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20tests/end2end/boost/experiment-controller-tests.js%3AL51-60%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-Pull ba4ee2925514270db1a37df0443cb78d2c29a2bd tests/end2end/boost/experiment-controller-tests.js 6'></a> &nbsp;<b><a href='https://github.com/Natural-Intelligence/rule-bos/pull/29#discussion_r91548723'>File: tests/end2end/boost/experiment-controller-tests.js:L51-60</a></b>
- <a href='https://github.com/moranzabatani'></a> why remove all others parameter from context?
- [ ] <a href='#crh-comment-Pull ba4ee2925514270db1a37df0443cb78d2c29a2bd tests/end2end/umbrella/rule-priority-controller-tests.js 56'></a> &nbsp;<b><a href='https://github.com/Natural-Intelligence/rule-bos/pull/29#discussion_r91549544'>File: tests/end2end/umbrella/rule-priority-controller-tests.js:L1-316</a></b>
- <a href='https://github.com/moranzabatani'></a> Need to add e2e tests (involves rule & rule priority unit tests) that:
1. create a rule
2. create a rule prioirity
3. create a rule
4. update rule prioirity
5. verify order
6. delete rule
7. update priority
8. verify order
Etc.
- <a href='https://github.com/igor-lu'></a> rules and rules priorities are fully independent in server side
- [ ] <a href='#crh-comment-Pull ba4ee2925514270db1a37df0443cb78d2c29a2bd tests/end2end/umbrella/rule-priority-controller-tests.js 83'></a> &nbsp;<b><a href='https://github.com/Natural-Intelligence/rule-bos/pull/29#discussion_r91549716'>File: tests/end2end/umbrella/rule-priority-controller-tests.js:L1-316</a></b>
- <a href='https://github.com/moranzabatani'></a> Need to add e2e tests (involves experiment & experiment priority unit tests) - the same as above

<a href='https://www.codereviewhub.com/Natural-Intelligence/rule-bos/pull/29?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/Natural-Intelligence/rule-bos/pull/29?approve=1'></a>&nbsp;<a href='https://github.com/Natural-Intelligence/rule-bos/pull/29'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by omerh** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/Natural-Intelligence/rule-bos/pull/29'>Refresh</a>.<a href='#crh-update-none'></a>

File: lib/controllers/experiment-priority-controller.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+'use strict';
+const infra = require('ni-node-infra');
+const EXPERIMENTS_PRIORITIES_REST_BASE_PATH = '/api/v1/experiment-priorities/';
+const platform = require('./../platform-helper');
+
+const controller = module.exports.controller = function controller(app) {
+  const experimentPriorityService = require('./../services/experiment-priority-service');
+
+  const queryableFields = platform
+    .boost(['type', 'context.siteId'])
```
Comment (moranzabatani) [inline lib/controllers/experiment-priority-controller.js]: Should be boostSiteId

File: lib/controllers/experiment-priority-controller.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    .umb(['type', 'context.segmentId', 'context.siteId', 'context.verticalId']);
+
+  /**
+   * @type {{
+   *     addAllCRUD()
+   * }}
+   */
+  const crudController = new infra.CrudController(app, EXPERIMENTS_PRIORITIES_REST_BASE_PATH, experimentPriorityService, {queryableFields});
+  // remove unnecessary api (publish)
+  crudController.addAllCRUD();
```
Comment (moranzabatani) [inline lib/controllers/experiment-priority-controller.js]: We do not support delete and publish api, so i suggest to remove so no one will use it by mistake.

File: lib/controllers/rule-priority-controller.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+'use strict';
+const infra = require('ni-node-infra');
+const RULES_PRIORITIES_REST_BASE_PATH = '/api/v1/rule-priorities/';
+const platform = require('./../platform-helper');
+
+const controller = module.exports.controller = function controller(app) {
+  const rulePriorityService = require('./../services/rule-priority-service');
+
+  const queryableFields = platform
+    .boost(['type', 'context.siteId'])
```
Comment (moranzabatani) [inline lib/controllers/rule-priority-controller.js]: Should be boostSiteId

File: lib/controllers/rule-priority-controller.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+  /**
+   * @type {{
+   *     addAllCRUD()
+   * }}
+   */
+  const crudController = new infra.CrudController(app, RULES_PRIORITIES_REST_BASE_PATH, rulePriorityService, {queryableFields});
+  // remove unnecessary api (publish)
+  crudController.addAllCRUD();
+};
```
Comment (moranzabatani) [inline lib/controllers/rule-priority-controller.js]: The same here - remove publish & delete endpoints.

File: lib/repositories/experiment-repository.js [hunk 5]
```diff
@@ -26,6 +26,7 @@ const connectionHelper = core.dbConnectionHelper.instance;
 const connection = connectionHelper.getConnection('backOfficeDb');
 const experimentsModel = connection.models.experiments;
 
+//todo remove it (no priority in rules or in tests)
```
Comment (moranzabatani) [inline lib/repositories/experiment-repository.js]: when we need to remove it?

File: lib/services/experiment-priority-service.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+  /**
+   * ServiceResult
+   */
+  const existEntities = yield service.getEntities({
+    queryParams
+  });
+
+  if (!existEntities.ok) {
+    errors.push(errorModel('context', 'Failed on unique test'));
```
Comment (moranzabatani) [inline lib/services/experiment-priority-service.js]: Change error description to:Failed to verify uniqueness of Priority context

File: tests/end2end/umbrella/rule-priority-controller-tests.js [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+const collectionName = 'rule-priorities';
+
+const connectionHelper = core.dbConnectionHelper.instance;
+const rulePriorityModel = connectionHelper.getConnection('backOfficeDb').models['rule-priorities'];
+const previewRulesModel = connectionHelper.getConnection('previewDb').models['rule-priorities'];
+
+const previewRepository = new RtDbRepo(collectionName, niEntities.rt.rulePriority, 'previewDb');
+const publishRepository = new RtDbRepo(collectionName, niEntities.rt.rulePriority, 'publishDb');
+
+describe('RulePriority priority Controller class', () => {
```
Comment (moranzabatani) [inline tests/end2end/umbrella/rule-priority-controller-tests.js]: Need to add e2e tests (involves rule & rule priority unit tests) that:
1. create a rule
2. create a rule prioirity
3. create a rule
4. update rule prioirity
5. verify order
6. delete rule
7. update priority
8. verify order

Etc.

File: tests/end2end/umbrella/rule-priority-controller-tests.js [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+  });
+
+  after(() => {
+    restorePlatform();
+  });
+
+  beforeEach(cleanup);
+  afterEach(cleanup);
+
+  function cleanup(done) {
```
Comment (moranzabatani) [inline tests/end2end/umbrella/rule-priority-controller-tests.js]: Need to add e2e tests (involves experiment & experiment priority unit tests) - the same as above

File: tests/end2end/umbrella/rule-priority-controller-tests.js (hunk 7 above)
Comment (igor-lu) [inline tests/end2end/umbrella/rule-priority-controller-tests.js]: rules and rules priorities are fully independent in server side

**Pull request merged**